transport/httptransport/libp2p_server.go

```diff
@@ -0,0 +1,493 @@
+		AuthToken:    authToken,
+	}
+	return dt, nil
+}
+
+// CleanupPreparedRequest is called when a request cannot be fulfilled,
+// for example because the provider rejected the deal proposal
+func (s *Libp2pCarServer) CleanupPreparedRequest(ctx context.Context, authToken string, proposalCid cid.Cid) error {
+	// Clean up any transfer that was created
+	s.transfersLk.Lock()
+	xfer, ok := s.transfers[proposalCid]
+	if ok {
+		delete(s.transfers, proposalCid)
+		xfer.Cancel(fmt.Errorf("cancelled"))
+	}
+	s.transfersLk.Unlock()
+
+	// Delete the auth token for the request
+	return s.auth.Delete(ctx, authToken)
+}
+
+func (s *Libp2pCarServer) handler(w http.ResponseWriter, r *http.Request) {
+	authToken, authVal, herr := s.checkAuth(r)
+	if herr != nil {
+		log.Infow("data transfer request failed", "code", herr.code, "err", herr.error)
+		w.WriteHeader(herr.code)
+		return
+	}
+
+	// Get a block info cache for the CarOffsetWriter
+	bic := s.bicm.Get(authVal.PayloadCid)
+	err := s.serveContent(w, r, authToken, authVal, bic)
+	s.bicm.Unref(authVal.PayloadCid, err)
+}
+
+func (s *Libp2pCarServer) checkAuth(r *http.Request) (string, *authValue, *httpError) {
+	ctx := r.Context()
+
+	// Get auth token from Authorization header
+	_, authToken, ok := r.BasicAuth()
+	if !ok {
+		return "", nil, &httpError{
+			error: errors.New("rejected request with no Authorization header"),
+			code:  401,
+		}
+	}
+
+	// Get proposal CID from auth datastore
+	authValueJson, err := s.auth.Get(ctx, authToken)
+	if xerrors.Is(err, ErrTokenNotFound) {
+		return "", nil, &httpError{
+			error: errors.New("rejected unrecognized auth token"),
+			code:  401,
+		}
+	} else if err != nil {
+		return "", nil, &httpError{
+			error: fmt.Errorf("getting key from datastore: %w", err),
+			code:  500,
+		}
+	}
+
+	var val authValue
+	err = json.Unmarshal(authValueJson, &val)
+	if err != nil {
+		return "", nil, &httpError{
+			error: fmt.Errorf("unmarshaling json from datastore: %w", err),
+			code:  500,
+		}
+	}
+
+	return authToken, &val, nil
+}
+
+func (s *Libp2pCarServer) serveContent(w http.ResponseWriter, r *http.Request, authToken string, val *authValue, bic *car.BlockInfoCache) error {
+	ctx := r.Context()
+
+	// Create a CarOffsetWriter and a reader for it
+	cow := car.NewCarOffsetWriter(val.PayloadCid, s.bstore, bic)
+	content := car.NewCarReaderSeeker(ctx, cow, val.Size)
+
+	// Set the Content-Type header explicitly so that http.ServeContent doesn't
+	// try to do it implicitly
+	w.Header().Set("Content-Type", "application/car")
+
+	if r.Method == "HEAD" {
+		// For an HTTP HEAD request we don't send any data (just headers)
+		http.ServeContent(w, r, "", time.Time{}, content)
+
+		return nil
+	}
+
+	// Send the CAR file
+	return s.sendCar(r, w, val, authToken, content)
+}
+
+func (s *Libp2pCarServer) sendCar(r *http.Request, w http.ResponseWriter, val *authValue, authToken string, content *car.CarReaderSeeker) error {
+	// Create a new transfer
+	s.transfersLk.Lock()
```

Comment:
Taking this lock here will block the transfer if another thread has called `CleanupPreparedRequest` as that method also takes the `transfersLk ` and then blocks on closing the corresponding CarReadSeeker which in turn will have to wait for the read from that CARReadSeeker to finish.

Essentially, a new transfer can get delayed/blocked because of cancelling an ongoing transfer.

As discussed, a better way to do this is to have `sendCar` here emit events about the entire transfer state which are then consumed async by another component which then

a. Sticks the state in a map so the external world can query it.
b. Forwards the transfer state notifications to the external world.

Then, `sendCar` here does NOT need to take this transfer lock that contends with cancelling another transfer.

Reply:
Fixed